### PR TITLE
`azurerm_service_fabric_cluster` - support for specifying the cluster code version

### LINKED ISSUE #1906: azurerm_service_fabric_cluster upgrade_mode manual should be able to specify cluster_code_version
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Currently we allow either 'Automatic' or 'Manual' for upgrade_mode of Service Fabric Clusters, when selecting 'Manual' we are left to the latest supported version.  It would be preferable if the version could be specified.

You could argue that with the rate at which Microsoft drops support for versions would mean latest would be acceptable, but realistically we would want to control the version of Service Fabric we are using.

### New or Affected Resource(s)

* azurerm_service_fabric_cluster

### Potential Terraform Configuration

```hcl
resource "azurerm_service_fabric_cluster" "service_fabric" {
  name                = "test"
  resource_group_name = "${azurerm_resource_group.test.name}"
  location            = "${azurerm_resource_group.test.location}"
  reliability_level   = "Bronze"
  vm_image            = "Windows"
  management_endpoint = "https://test.cloudapp.azure.com:19080"

  node_type {
    name                 = "backend"
    instance_count       = 5
    is_primary           = true
    client_endpoint_port = 19000
    http_endpoint_port   = 19080
  }

  upgrade_mode         = "Manual"
  cluster_code_version = "6.3.176.9494"
}
```

### PR DESCRIPTION
**Pull request by steve-hawkins**
this should resolve #1906

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# make sure all .go files stay LF, this will help developers on Windows
```
Comment (katbyte) [inline .gitattributes]: This file is outside of the scope of this PR.

File: azurerm/resource_arm_service_fabric_cluster.go [hunk 2]
```diff
@@ -53,6 +53,11 @@ func resourceArmServiceFabricCluster() *schema.Resource {
 				}, false),
 			},
 
+			"cluster_code_version": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_service_fabric_cluster.go]: Could we get some validation here? Regex to enfore a `x.x.x.x` version would be nice but a `validation.NoZeroValues` is better then nothing

File: azurerm/resource_arm_service_fabric_cluster_test.go [hunk 3]
```diff
@@ -432,6 +467,34 @@ resource "azurerm_service_fabric_cluster" "test" {
 `, rInt, location, rInt, count)
 }
 
+func testAccAzureRMServiceFabricCluster_manualClusterCodeVersion(rInt int, location, clusterCodeVersion string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name = "acctestRG-%[1]d"
```
Comment (katbyte) [inline azurerm/resource_arm_service_fabric_cluster_test.go]: Very minor but could we align these assignments?

Comment (steve-hawkins) [reply azurerm/resource_arm_service_fabric_cluster_test.go]: sorry copy and paste job

File: azurerm/resource_arm_service_fabric_cluster.go [hunk 4]
```diff
@@ -423,6 +433,7 @@ func resourceArmServiceFabricClusterRead(d *schema.ResourceData, meta interface{
 		d.Set("management_endpoint", props.ManagementEndpoint)
 		d.Set("reliability_level", string(props.ReliabilityLevel))
 		d.Set("upgrade_mode", string(props.UpgradeMode))
+		d.Set("cluster_code_version ", props.ClusterCodeVersion)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_service_fabric_cluster.go]: I believe the test failure here is because there's a space at the end of this string (so the `d.Set` can't find a matching property in the schema called `cluster_code_version ` [note the space] - so the Set fails)?

Comment (tombuildsstuff) [reply azurerm/resource_arm_service_fabric_cluster.go]: @steve-hawkins since this PR otherwise LGTM - I hope you don't mind but I've pushed a commit to remove the extra space (and will kick off the tests now)

File: azurerm/resource_arm_service_fabric_cluster.go [hunk 5]
```diff
@@ -372,6 +381,7 @@ func resourceArmServiceFabricClusterUpdate(d *schema.ResourceData, meta interfac
 			NodeTypes:                    nodeTypes,
 			ReliabilityLevel:             servicefabric.ReliabilityLevel1(reliabilityLevel),
 			UpgradeMode:                  servicefabric.UpgradeMode1(upgradeMode),
+			ClusterCodeVersion:           utils.String(clusterCodeVersion),
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_service_fabric_cluster.go]: we've run the test suite here and this fails because the `ClusterCodeVersion` is an empty string when it's not specified (so this fails on the API Validation):

```
* azurerm_service_fabric_cluster.test: Error creating Service Fabric Cluster "acctest-9043244035322992792" (Resource Group "acctestRG-9043244035322992792"): servicefabric.ClustersClient#Create: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidRequestFormat" Message="Cannot parse the request." Details=[{"code":"InvalidRequestFormat","message":"Error parsing version string:. Path 'properties.clusterCodeVersion', line 1, position 115."}]
```

could we update this to only set the field if it's got a value? e.g.
```
if clusterCodeVersion != "" {
  parameters.ClusterPropertiesUpdateParameters.ClusterCodeVersion = utils.String(clusterCodeVersion)
}
```

(we also need to update the Create above to do the same thing)

**Comment by steve-hawkins** [comment]
Hi @tombuildsstuff, I'm having some trouble with the conditional logic on ClusterCodeVersion in the resourceArmServiceFabricClusterRead function.  I have all the current acceptance tests working with the following:-

```go
upgradeMode := string(props.UpgradeMode)
d.Set("upgrade_mode", upgradeMode)

if upgradeMode == "Manual" {
  d.Set("cluster_code_version", props.ClusterCodeVersion)
}
```

This however will fail if UpgradeMode is Manual and no version is specified:-

```
$ make fmt; make testacc TEST=./azurerm TESTARGS=-test.run=TestAccAzureRMServiceFabricCluster_manualUpgrade
gofmt -w $(find . -name '*.go' |grep -v vendor)
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./azurerm -v -test.run=TestAccAzureRMServiceFabricCluster_manualUpgrade -timeout 180m
=== RUN   TestAccAzureRMServiceFabricCluster_manualUpgrade
--- FAIL: TestAccAzureRMServiceFabricCluster_manualUpgrade (53.41s)
    testing.go:513: Step 0 error: After applying this step, the plan was not empty:

        DIFF:

        UPDATE: azurerm_service_fabric_cluster.test
          cluster_code_version: "6.3.176.9494" => ""

        STATE:

        azurerm_resource_group.test:
          ID = /subscriptions/020ef497-e33d-4705-859b-08663256a4bc/resourceGroups/acctestRG-8531727941990102137
          provider = provider.azurerm
          location = uksouth
          name = acctestRG-8531727941990102137
          tags.% = 0
        azurerm_service_fabric_cluster.test:
          ID = /subscriptions/020ef497-e33d-4705-859b-08663256a4bc/resourcegroups/acctestRG-8531727941990102137/providers/Microsoft.ServiceFabric/clusters/acctest-8531727941990102137
          provider = provider.azurerm
          add_on_features.# = 0
          certificate.# = 0
          client_certificate_thumbprint.# = 0
          cluster_code_version = 6.3.176.9494
          cluster_endpoint = https://uksouth.servicefabric.azure.com/runtime/clusters/64d87576-e429-4e0e-ac9e-db8b20b4f546
          diagnostics_config.# = 0
          fabric_settings.# = 0
          location = uksouth
          management_endpoint = http://example:80
          name = acctest-8531727941990102137
          node_type.# = 1
          node_type.0.application_ports.# = 1
          node_type.0.application_ports.0.end_port = 30000
          node_type.0.application_ports.0.start_port = 20000
          node_type.0.client_endpoint_port = 2020
          node_type.0.durability_level = Bronze
          node_type.0.ephemeral_ports.# = 1
          node_type.0.ephemeral_ports.0.end_port = 65534
          node_type.0.ephemeral_ports.0.start_port = 49152
          node_type.0.http_endpoint_port = 80
          node_type.0.instance_count = 3
          node_type.0.is_primary = true
          node_type.0.name = first
          reliability_level = Bronze
          resource_group_name = acctestRG-8531727941990102137
          tags.% = 0
          upgrade_mode = Manual
          vm_image = Windows

          Dependencies:
            azurerm_resource_group.test
FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm       54.873s
make: *** [testacc] Error 1
```

I could make it so the cluster_code_version has to be specified if the upgrade_mode is set to Manual, would that be acceptable?

I might need some help if we want this to work with upgrade_mode to Manual and no cluster_code_version specified

**Comment by steve-hawkins** [comment]
Hi @tombuildsstuff and @katbyte , I'm having some trouble with the conditional logic on ClusterCodeVersion in the resourceArmServiceFabricClusterRead function.  I have all the current acceptance tests working with the following:-

```go
upgradeMode := string(props.UpgradeMode)
d.Set("upgrade_mode", upgradeMode)

if upgradeMode == "Manual" {
  d.Set("cluster_code_version", props.ClusterCodeVersion)
}
```

This however will fail if UpgradeMode is Manual and no version is specified:-

```
$ make fmt; make testacc TEST=./azurerm TESTARGS=-test.run=TestAccAzureRMServiceFabricCluster_manualLatest
gofmt -w $(find . -name '*.go' |grep -v vendor)
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./azurerm -v -test.run=TestAccAzureRMServiceFabricCluster_manualLatest -timeout 180m
=== RUN   TestAccAzureRMServiceFabricCluster_manualLatest
--- FAIL: TestAccAzureRMServiceFabricCluster_manualLatest (53.41s)
    testing.go:513: Step 0 error: After applying this step, the plan was not empty:

        DIFF:

        UPDATE: azurerm_service_fabric_cluster.test
          cluster_code_version: "6.3.176.9494" => ""

        STATE:

        azurerm_resource_group.test:
          ID = /subscriptions/[REDACTED]/resourceGroups/acctestRG-8531727941990102137
          provider = provider.azurerm
          location = uksouth
          name = acctestRG-8531727941990102137
          tags.% = 0
        azurerm_service_fabric_cluster.test:
          ID = /subscriptions/[REDACTED]/resourcegroups/acctestRG-8531727941990102137/providers/Microsoft.ServiceFabric/clusters/acctest-8531727941990102137
          provider = provider.azurerm
          add_on_features.# = 0
          certificate.# = 0
          client_certificate_thumbprint.# = 0
          cluster_code_version = 6.3.176.9494
          cluster_endpoint = https://uksouth.servicefabric.azure.com/runtime/clusters/64d87576-e429-4e0e-ac9e-db8b20b4f546
          diagnostics_config.# = 0
          fabric_settings.# = 0
          location = uksouth
          management_endpoint = http://example:80
          name = acctest-8531727941990102137
          node_type.# = 1
          node_type.0.application_ports.# = 1
          node_type.0.application_ports.0.end_port = 30000
          node_type.0.application_ports.0.start_port = 20000
          node_type.0.client_endpoint_port = 2020
          node_type.0.durability_level = Bronze
          node_type.0.ephemeral_ports.# = 1
          node_type.0.ephemeral_ports.0.end_port = 65534
          node_type.0.ephemeral_ports.0.start_port = 49152
          node_type.0.http_endpoint_port = 80
          node_type.0.instance_count = 3
          node_type.0.is_primary = true
          node_type.0.name = first
          reliability_level = Bronze
          resource_group_name = acctestRG-8531727941990102137
          tags.% = 0
          upgrade_mode = Manual
          vm_image = Windows

          Dependencies:
            azurerm_resource_group.test
FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm       54.873s
make: *** [testacc] Error 1
```

I could make it so the cluster_code_version has to be specified if the upgrade_mode is set to Manual, would that be acceptable?

I might need some help if we want this to work with upgrade_mode to Manual and no cluster_code_version specified

**Comment by katbyte** [comment]
@steve-hawkins,

Might i suggest always reading in `cluster_code_version`:

```go
  d.Set("cluster_code_version", v) //d.Set will handle nil values
```

and then adding computed to the `cluster_code_version` property so terraform expects it to be read in

**Comment by tombuildsstuff** [comment]
Tests pass:

screenshot 2018-09-24 at 11 58 59

**Comment by steve-hawkins** [comment]
Thanks @katbyte @tombuildsstuff I knew I would have been missing something obvious

Much appreciated and looking forward to contributing more

**Comment by steve-hawkins** [comment]
@tombuildsstuff any chance this could make it into 1.16.0?
